Pandas Assignment...

Data Analisys with Pandas 

SF Salaries 
Welcome to a quick exercise for you to practice your pandas skills! We will be using the SF Salaries Dataset Just follow along and complete the tasks outlined in bold below. The tasks will get harder and harder as you go along.

Q1:What is the average BasePay ?
A: 66325.44884050643

In [1]:
# Import the pandas library with the alias 'pd'.
import pandas as pd

# Read the data from the CSV file "Salaries.csv" into a DataFrame 'df'.
df = pd.read_csv("Salaries.csv")

# Display the first few rows of the DataFrame 'df' using the 'head()' method.
df.head()


Id       EmployeeName                                        JobTitle  \
0   1     NATHANIEL FORD  GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY   
1   2       GARY JIMENEZ                 CAPTAIN III (POLICE DEPARTMENT)   
2   3     ALBERT PARDINI                 CAPTAIN III (POLICE DEPARTMENT)   
3   4  CHRISTOPHER CHONG            WIRE ROPE CABLE MAINTENANCE MECHANIC   
4   5    PATRICK GARDNER    DEPUTY CHIEF OF DEPARTMENT,(FIRE DEPARTMENT)   

     BasePay  OvertimePay   OtherPay  Benefits   TotalPay  TotalPayBenefits  \
0  167411.18         0.00  400184.25       NaN  567595.43         567595.43   
1  155966.02    245131.88  137811.38       NaN  538909.28         538909.28   
2  212739.13    106088.18   16452.60       NaN  335279.91         335279.91   
3   77916.00     56120.71  198306.90       NaN  332343.61         332343.61   
4  134401.60      9737.00  182234.59       NaN  326373.19         326373.19   

   Year  Notes         Agency  Status  
0  2011    NaN  San Francisco     NaN  
1  2011    NaN  San Francisco     NaN  
2  2011    NaN  San Francisco     NaN  
3  2011    NaN  San Francisco     NaN  
4  2011    NaN  San Francisco     NaN

In [9]:
# Calculate the mean (average) of the 'BasePay' column in the DataFrame 'df'.
average_base_pay = df['BasePay'].mean()
print(average_base_pay)

66325.4488404877


Q2: What is the highest amount of OvertimePay in the dataset ?
A:245131.88

In [10]:
# Find the maximum value in the 'OvertimePay' column of the DataFrame 'df'.
max_overtime_pay = df['OvertimePay'].max()
print(max_overtime_pay)

245131.88


Q3:What is the job title of JOSEPH DRISCOLL ? Note: Use all caps, otherwise you may get an answer that doesn't match up (there is also a lowercase Joseph Driscoll). 
A: 24    CAPTAIN, FIRE SUPPRESSION

In [4]:
# Filter the DataFrame 'df' to get the 'JobTitle' of the employee with the name 'JOSEPH DRISCOLL'.
name = df[df['EmployeeName'] == 'JOSEPH DRISCOLL']['JobTitle']

# Print the 'JobTitle' for the employee with the specified name.
print(name)

# job_title_value = name['JobTitle'].values[0]


24    CAPTAIN, FIRE SUPPRESSION
Name: JobTitle, dtype: object


Q4: How much does JOSEPH DRISCOLL make (including benefits)?
A: 24    270324.91

In [5]:
# Filter the DataFrame 'df' to get the 'TotalPayBenefits' of the employee with the name 'JOSEPH DRISCOLL'.
total_pay_benefits = df[df['EmployeeName'] == 'JOSEPH DRISCOLL']['TotalPayBenefits']

# Print the 'TotalPayBenefits' for the employee with the specified name.
print(total_pay_benefits)


24    270324.91
Name: TotalPayBenefits, dtype: float64


Q5: What is the name of highest paid person (including benefits)?

In [6]:
# Find the maximum value in the 'TotalPayBenefits' column.
max_total_pay_benefits = df['TotalPayBenefits'].max()

# Filter the DataFrame 'df' to get the row(s) where 'TotalPayBenefits' is equal to the maximum value.
max_total_pay_benefits_row = df[df['TotalPayBenefits'] == max_total_pay_benefits]

# Print the 'EmployeeName' from the filtered row(s).
print(max_total_pay_benefits_row['EmployeeName'])


0    NATHANIEL FORD
Name: EmployeeName, dtype: object


Q6: What is the name of lowest paid person (including benefits)? Do you notice something strange about how much he or she is paid? 
Note: with full details
A: 148653


In [7]:
# Find the minimum value in the 'TotalPayBenefits' column.
min_total_pay_benefits = df['TotalPayBenefits'].min()

# Filter the DataFrame 'df' to get the row(s) where 'TotalPayBenefits' is equal to the minimum value.
min_total_pay_benefits_row = df[df['TotalPayBenefits'] == min_total_pay_benefits]

# Print the details of the row(s) where 'TotalPayBenefits' is at its minimum.
print(min_total_pay_benefits_row)

# Print the name of the lowest-paid person from the filtered row(s).
print(f"\nLowest paid person is: {min_total_pay_benefits_row['EmployeeName'].values[0]}")

# Provide additional information about the observation.
print("The strange thing noticed is basepay, overtimepay is 0, but benefits are in negative.")


            Id EmployeeName                    JobTitle  BasePay  OvertimePay  \
148653  148654    Joe Lopez  Counselor, Log Cabin Ranch      0.0          0.0   

        OtherPay  Benefits  TotalPay  TotalPayBenefits  Year  Notes  \
148653   -618.13       0.0   -618.13           -618.13  2014    NaN   

               Agency  Status  
148653  San Francisco     NaN  

Lowest paid person is: Joe Lopez
The strange thing noticed is basepay, overtimepay is 0, but benefits are in negative.


Q7:What was the average (mean) BasePay of all employees per year? (2011-2014) ? 
A:  Year
2011    63595.956517
2012    65436.406857
2013    69630.030216
2014    66564.421924
Name: BasePay, dtype: float64

In [11]:
# Group the DataFrame 'df' by the 'Year' column.
grouped_by_year = df.groupby('Year')

# Print the data type of the 'Year' column.
print(type(df.Year))

# Calculate the mean of 'BasePay' for each group (each year).
average_base_pay_by_year = grouped_by_year['BasePay'].mean()

# Print the result.
print(average_base_pay_by_year)



<class 'pandas.core.series.Series'>
Year
2011    63595.956517
2012    65436.406857
2013    69630.030216
2014    66564.421924
Name: BasePay, dtype: float64


Q8: How many unique job titles are there?
A:2159

In [ ]:
# df['JobTitle'].nunique()

# Count the number of unique job titles in the 'JobTitle' column of the DataFrame 'df'.
unique_job_titles_count = df['JobTitle'].nunique()


2159

Q9:What are the top 5 most common jobs?
A: Transit Operator                7036
Special Nurse                   4389
Registered Nurse                3736
Public Svc Aide-Public Works    2518
Police Officer 3                2421

In [ ]:
# df.JobTitle.value_counts().head(5)

# Count the occurrences of each unique job title in the 'JobTitle' column of the DataFrame 'df'.
title_counts = df['JobTitle'].value_counts()

# Select the top 5 most common job titles based on their occurrence counts.
top_5_titles = title_counts.head(5)


JobTitle
Transit Operator                7036
Special Nurse                   4389
Registered Nurse                3736
Public Svc Aide-Public Works    2518
Police Officer 3                2421
Name: count, dtype: int64

Q10:How many Job Titles were represented by only one person in 2013? (e.g. Job Titles with only one occurence in 2013?)
A: 202

In [ ]:
# Filter the DataFrame 'df' to include only rows where the 'Year' column is equal to 2013.
# Select the 'JobTitle' column for these filtered rows.
filtered_titles_2013 = df[df['Year'] == 2013]['JobTitle']

# Count the occurrences of each unique job title in the filtered 'JobTitle' column.
title_counts_2013 = filtered_titles_2013.value_counts()

# Sum the number of titles that have a count of 1, indicating titles with only one occurrence in 2013.
count_single_occurrence_titles_2013 = sum(title_counts_2013 == 1)


202

Q11:How many people have the word Chief in their job title? (This is pretty tricky)
A:627

In [ ]:
# Define a function named 'chief' that takes a 'title' parameter.
def chief(title):
    # Check if the lowercase version of 'title' contains the substring 'chief'.
    if 'chief' in title.lower():
        # Return True if 'chief' is found, indicating that the title contains 'chief'.
        return True
    else:
        # Return False if 'chief' is not found.
        return False

# Apply the 'chief' function to each element in the 'JobTitle' column of the DataFrame 'df'.
# Sum the resulting True values, representing the count of titles containing 'chief'.
count_chief_titles = sum(df['JobTitle'].apply(chief))


627

Q12: Bonus: Is there a correlation between length of the Job Title string and Salary? 
A: 
			TotalPayBenefits	title_len
TotalPayBenefits	1.000000		-0.036878
title_len		-0.036878		1.000000

In [ ]:
# Creating a new column 'titlelen' in the DataFrame 'df', which contains the length of each 'JobTitle'.
df['titlelen'] = df['JobTitle'].apply(len)

# Creating a new DataFrame that contains two columns: 'titlelen' and 'TotalPayBenefits'.
# Calculating the correlation between the two columns.
correlation_result = df[['titlelen', 'TotalPayBenefits']].corr()


titlelen  TotalPayBenefits
titlelen          1.000000         -0.036878
TotalPayBenefits -0.036878          1.000000